src/event/cmap.rs

```diff
@@ -327,3 +335,100 @@ pub trait CmapEventHandler: Send + Sync {
     /// whenever a connection is checked back into a connection pool.
     fn handle_connection_checked_in_event(&self, _event: ConnectionCheckedInEvent) {}
 }
+
+#[derive(Clone)]
+pub(crate) enum CmapEvent {
```

Comment:
I wasn't sure if it was worth trying to combine this with the `cmap::test::event::Event` type in the tests because that has a decent amount very Astrolabe-specific serialization logic associated with it which I think we'd have to pull up into the driver itself. Let me know if you think that refactor makes sense though.

Reply:
From looking at these [docs](https://doc.rust-lang.org/reference/conditional-compilation.html#the-cfg_attr-attribute), I think that it would be possible to do something like `#[cfg_attr(test, derive(Deserialize)]` on this enum to define a `Deserialize` implementation conditionally, and then do the same for the custom deserialization attributes. Then we could avoid doing a custom implementation that requires the duplicate struct.

What was your concern about having the `Deserialize` impl on this enum in the regular driver code, though? Because it's `pub(crate)` it shouldn't leak into the public API, but maybe I'm missing another issue at hand.

Reply:
I think I'm maybe trying to follow more of a Swift driver precendent here where we tended to, when possible, put any Codable implementations that were custom and specific to test file structure in the tests rather than in the driver itself. (was not always possible if the only change was custom `CodingKeys`, but can be done for anything with custom `init(from: Decoder)` and encode(to: Encoder)` methods). 

in this case I think we would need to pull up all of this: https://github.com/mongodb/mongo-rust-driver/blob/5ad1c6127138ecd26b95256ac98ac937fab7f220/src/cmap/test/event.rs#L214-L301 

but if that's not a concern and/or we have existing precedent for this type of thing being in the driver itself then I don't mind moving the logic

Reply:
If we have to pull in all that test code, then I think it makes sense to go with separate implementations. The risk is low, but less code in the release driver -> fewer bug opportunities.

>I think I'm maybe trying to follow more of a Swift driver precendent here where we tended to, when possible, put any Codable implementations that were custom and specific to test file structure in the tests rather than in the driver itself.

We do have `Serialize` and `Deserialize` implementations that are derived on public types but used only for tests. I think we treat that kind of thing as a "SHOULD" for convenience's sake. In this case, there's a lot more custom logic than a simple derive and you've already done the "SHOULD" part, so it seems fine to keep it separate imo.